juniper_actix/src/lib.rs

```diff
@@ -217,6 +217,257 @@ pub async fn playground_handler(
+                            .await
+                            .expect("Infallible: this should not happen");
+                    }
+                    .into_actor(self)
+                    .wait(ctx);
+                }
+                Err(_) => {
+                    // ignore the message if there's a transport error
+                    // TODO: log the event when tracing is merged
+                }
+            }
+        }
+    }
+
+    /// juniper -> actor
+    impl<Query, Mutation, Subscription, CtxT, S, I> Actor
+        for SubscriptionActor<Query, Mutation, Subscription, CtxT, S, I>
+    where
+        Query: GraphQLTypeAsync<S, Context = CtxT> + Send + 'static,
+        Query::TypeInfo: Send + Sync,
+        Mutation: GraphQLTypeAsync<S, Context = CtxT> + Send + 'static,
+        Mutation::TypeInfo: Send + Sync,
+        Subscription: GraphQLSubscriptionType<S, Context = CtxT> + Send + 'static,
+        Subscription::TypeInfo: Send + Sync,
+        CtxT: Unpin + Send + Sync + 'static,
+        S: ScalarValue + Send + Sync + 'static,
+        I: Init<S, CtxT> + Send,
+    {
+        type Context = ws::WebsocketContext<Self>;
+
+        fn started(&mut self, ctx: &mut Self::Context) {
+            let stream = self.graphql_rx.clone();
+            let addr = ctx.address();
+
+            let fut = async move {
+                let mut stream = stream.lock().await;
+                while let Some(message) = stream.next().await {
+                    // sending the message to self so that it can be forwarded back to the client
+                    addr.do_send(ServerMessageWrapper { message });
+                }
+            }
+            .into_actor(self);
+
+            ctx.spawn(fut);
+        }
+    }
+
+    /// actor -> websocket response
+    impl<Query, Mutation, Subscription, CtxT, S, I> Handler<ServerMessageWrapper<S>>
+        for SubscriptionActor<Query, Mutation, Subscription, CtxT, S, I>
+    where
+        Query: GraphQLTypeAsync<S, Context = CtxT> + Send + 'static,
+        Query::TypeInfo: Send + Sync,
+        Mutation: GraphQLTypeAsync<S, Context = CtxT> + Send + 'static,
+        Mutation::TypeInfo: Send + Sync,
+        Subscription: GraphQLSubscriptionType<S, Context = CtxT> + Send + 'static,
+        Subscription::TypeInfo: Send + Sync,
+        CtxT: Unpin + Send + Sync + 'static,
+        S: ScalarValue + Send + Sync + 'static,
+        I: Init<S, CtxT> + Send,
+    {
+        type Result = ();
+
+        fn handle(
+            &mut self,
+            msg: ServerMessageWrapper<S>,
+            ctx: &mut ws::WebsocketContext<Self>,
+        ) -> Self::Result {
+            let msg = serde_json::to_string(&msg.message);
+
+            match msg {
+                Ok(msg) => {
+                    ctx.text(msg);
+                }
+                Err(e) => {
+                    // TODO: log the event when tracing is merged
+                    ctx.text(format!("error serializing response: {}", e));
```

Comment:
I'm a bit inclined to say we should actually close the connection if this ever happens. It seems like it would generally be better to close the connection and let the client re-open it than leave the client waiting for a message that'll never arrive.

So maybe something like this would be appropriate:

```suggestion
                    ctx.close(ws::CloseCode::Error, format!("error serializing response: {}", e));
```

Reply:
Yep, I agree. I've just done it. It looks like sending the close frame also closes the connection and stops the actor 👍